system/include/emscripten/atomic.h

```diff
@@ -19,63 +19,202 @@ extern "C" {
 
 // Atomically stores the given value to the memory location, and returns the
 // value that was there prior to the store.
-uint8_t emscripten_atomic_exchange_u8(void/*uint8_t*/ *addr __attribute__((nonnull)), uint8_t newVal);
-uint16_t emscripten_atomic_exchange_u16(void/*uint16_t*/ *addr __attribute__((nonnull)), uint16_t newVal);
-uint32_t emscripten_atomic_exchange_u32(void/*uint32_t*/ *addr __attribute__((nonnull)), uint32_t newVal);
-uint64_t emscripten_atomic_exchange_u64(void/*uint64_t*/ *addr __attribute__((nonnull)), uint64_t newVal);
+static __attribute__((always_inline))
+uint8_t emscripten_atomic_exchange_u8(void /*uint8_t*/* addr __attribute__((nonnull)), uint8_t newVal) {
+  return __c11_atomic_exchange((_Atomic uint8_t*)addr, newVal, __ATOMIC_SEQ_CST);
+}
```

Comment:
Taking a peek now at e.g. emmintrin.h, which I modeled according to Clang system SSE headers, it uses the form

```c++
static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
_mm_cmpeq_pd(__m128d __a, __m128d __b)
{
  return (__m128d)wasm_f64x2_eq((v128_t)__a, (v128_t)__b);
}
```

Not sure why it adds both `__inline__ ` and `__always_inline__`, I'd think `__always_inline__` would be enough and `__inline__` would be redundant.

Though `__inline__` is used instead of `inline` so that the code would inline warnings-free even when an ancient `--std=c89` would be used, which doesn't have `inline` keyword.

Tried to search what the `__nodebug__` attribute does, but couldn't find a hit. I vaguely recall that might mean that debuggers should not step into the function, but should instead behave as if it is a single instruction.

I see in wasm_worker.h I had used just

```c++
static __attribute__((always_inline)) ATOMICS_WAIT_RESULT_T emscripten_wasm_wait_i32(int32_t *address __attribute__((nonnull)), int expectedValue, int64_t maxWaitNanoseconds)
```
which I now see is not consistent with the SSE headers, which I considered to be "the way Clang authors intended".

Maybe adding that `__inline__` and `__nodebug__` would be good everywhere if we want to be super precise, but works for me either way.

Reply:
Done